src/uk/gov/defra/ffc/Provision.groovy

```diff
@@ -90,41 +90,67 @@ class Provision implements Serializable {
     getResourceFile(ctx, resourcePath, filename, destinationFolder, true)
   }
 
-  private static def getMigrationEnvVars(ctx, environment, repoName, pr) {
-    def searchKeys = [
-      'POSTGRES_ADMIN_USERNAME',
-      'POSTGRES_ADMIN_PASSWORD',
-      'POSTGRES_HOST',
-      'POSTGRES_SCHEMA_PASSWORD'
-    ]
+  private static def getRepoPostgresEnvVars(ctx, environment, repoName, pr) {
     def appConfigPrefix = environment + '/'
-    def appConfigValues = Utils.getConfigValues(ctx, searchKeys, appConfigPrefix, Utils.defaultNullLabel, false)
-    appConfigValues['POSTGRES_ADMIN_PASSWORD'] = escapeQuotes(appConfigValues['POSTGRES_ADMIN_PASSWORD'])
-    appConfigValues['POSTGRES_SCHEMA_PASSWORD'] = escapeQuotes(appConfigValues['POSTGRES_SCHEMA_PASSWORD'])
+    def postgresUserKey = 'postgresService.postgresUser'
+    def postgresDbKey = 'postgresService.postgresDb'
+    
+    def appConfigValuesRepo = Utils.getConfigValues(ctx, [postgresUserKey, postgresDbKey], appConfigPrefix, repoName, false)
+
+    def schemaUser = appConfigValuesRepo[postgresUserKey]
+    if (!schemaUser) {
+      throw new Exception("No $postgresUserKey AppConfig for $repoName in $environment environment")
+    }
 
-    def migrationEnvVars = appConfigValues.collect { "$it.key=$it.value" }
+    def database = appConfigValuesRepo[postgresDbKey]
+    if (!database) {
+      throw new Exception("No $postgresDbKey AppConfig for $repoName in $environment environment")
+    }
 
+    def schemaRole = schemaUser.split('@')[0]
     def schemaName = repoName.replace('-','_') + pr
-    def schemaRole = "${schemaName}_role"
-    def schemaUser = getSchemaUserWithHostname(schemaRole, appConfigValues['POSTGRES_HOST'])
-    def databaseName = repoName.replace('-','_').replace('_service', '')
+    def token = getSchemaToken(ctx, schemaRole)
+
+    return [
+      "POSTGRES_DB=$database",
+      "POSTGRES_SCHEMA_NAME=$schemaName",
+      "POSTGRES_SCHEMA_ROLE=$schemaRole",
+      "POSTGRES_SCHEMA_USERNAME=$schemaUser",
+      "POSTGRES_SCHEMA_PASSWORD=$token"
+    ]
+  }
 
-    migrationEnvVars.add("POSTGRES_SCHEMA_ROLE=$schemaRole")
-    migrationEnvVars.add("POSTGRES_SCHEMA_USERNAME=$schemaUser")
-    migrationEnvVars.add("POSTGRES_SCHEMA_NAME=$schemaName")
-    migrationEnvVars.add("POSTGRES_DB=$databaseName")
+  private static def getSchemaToken(ctx, roleName) {
+    def clientId = ctx.sh(returnStdout: true, script: "az identity show --resource-group $ctx.AZURE_POSTGRES_RESOURCE_GROUP --name $roleName --query clientId --output tsv").trim()
+    return ctx.sh(returnStdout: true, script: "curl -s 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fossrdbms-aad.database.windows.net&client_id=$clientId' -H Metadata:true | jq -r .access_token").trim()
```

Comment:
Would it be worth adding this URL to an env var?

Reply:
I think this URL remains the same for all enviroments as it's an internal URL to distribute, i.e. https://serverfault.com/questions/427018/what-is-this-ip-address-169-254-169-254/427022

I think if we switch to using tokens for service bus, that might be a good time to split out the URL.